mmv1/third_party/terraform/data_sources/data_source_dns_keys.go

```diff
@@ -1,229 +1,436 @@
+	PublicKey    types.String `tfsdk:"public_key"`
+	Digests      types.List   `tfsdk:"digests"`
+
+	DSRecord types.String `tfsdk:"ds_record"`
+}
+
+type GoogleZoneSigningKeyDigest struct {
+	Digest types.String `tfsdk:"digest"`
+	Type   types.String `tfsdk:"type"`
+}
 
-		Schema: map[string]*schema.Schema{
-			"managed_zone": {
-				Type:             schema.TypeString,
-				Required:         true,
-				DiffSuppressFunc: compareSelfLinkOrResourceName,
+var (
+	digestAttrTypes = map[string]attr.Type{
+		"digest": types.StringType,
+		"type":   types.StringType,
+	}
+)
+
+func (d *GoogleDnsKeysDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_dns_keys"
+}
+
+func (d *GoogleDnsKeysDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		// This description is used by the documentation generator and the language server.
+		MarkdownDescription: "Get the DNSKEY and DS records of DNSSEC-signed managed zones",
+
+		Attributes: map[string]schema.Attribute{
+			"managed_zone": schema.StringAttribute{
+				Description:         "The Name of the zone.",
+				MarkdownDescription: "The Name of the zone.",
+				Required:            true,
 			},
-			"project": {
-				Type:     schema.TypeString,
-				Optional: true,
-				Computed: true,
-				ForceNew: true,
+			"project": schema.StringAttribute{
+				Description:         "The ID of the project for the Google Cloud.",
+				MarkdownDescription: "The ID of the project for the Google Cloud.",
+				Optional:            true,
+				Computed:            true,
 			},
-			"key_signing_keys": {
-				Type:     schema.TypeList,
-				Computed: true,
-				Elem:     kskResource(),
+			"id": schema.StringAttribute{
+				Description:         "DNS keys identifier",
+				MarkdownDescription: "DNS keys identifier",
+				Computed:            true,
 			},
-			"zone_signing_keys": {
-				Type:     schema.TypeList,
-				Computed: true,
-				Elem:     dnsKeyResource(),
+		},
+		Blocks: map[string]schema.Block{
+			"zone_signing_keys": schema.ListNestedBlock{
+				Description:         "A list of Zone-signing key (ZSK) records.",
+				MarkdownDescription: "A list of Zone-signing key (ZSK) records.",
+				NestedObject:        dnsKeyObject(),
+			},
+			"key_signing_keys": schema.ListNestedBlock{
+				Description:         "A list of Key-signing key (KSK) records.",
+				MarkdownDescription: "A list of Key-signing key (KSK) records.",
+				NestedObject:        kskObject(),
 			},
 		},
 	}
 }
 
-func dnsKeyResource() *schema.Resource {
-	return &schema.Resource{
-		Schema: map[string]*schema.Schema{
-			"algorithm": {
-				Type:     schema.TypeString,
+func (d *GoogleDnsKeysDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
```

Comment:
Per the DevEx team:
- During validation, the data source/resource Configure method is still called, even though the provider configuration doesn’t happen, because technically they can be configured with non-provider data too and we didn’t want to prevent folks from being able to do things like that in their code.
- If the provider configuration references other resource attributes that are unknown during planning, then the provider configuration has to choose whether it should error or return early without actually configuring anything (in the framework, this would mean things like `ConfigureRequest.ProviderData` being left empty)